src/passes/Table64Lowering.cpp

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2024 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// Lowers a module with a 64-bit table to one with a 32-bit table.
+//
+// This pass can be delete once table64 is implemented in the Wasm engines:
+// https://github.com/WebAssembly/memory64/issues/51
+//
+
+#include "ir/bits.h"
+#include "ir/import-utils.h"
+#include "pass.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+
+namespace wasm {
+
+static Name TABLE_BASE("__table_base");
+static Name TABLE_BASE32("__table_base32");
+
+struct Table64Lowering : public WalkerPass<PostWalker<Table64Lowering>> {
+
+  void wrapAddress64(Expression*& ptr, Name tableName) {
+    if (ptr->type == Type::unreachable) {
+      return;
+    }
+    auto& module = *getModule();
+    auto table = module.getTable(tableName);
+    if (table->is64()) {
+      assert(ptr->type == Type::i64);
+      Builder builder(module);
+      ptr = builder.makeUnary(UnaryOp::WrapInt64, ptr);
+    }
+  }
+
+  void extendAddress64(Expression*& ptr, Name tableName) {
+    if (ptr->type == Type::unreachable) {
+      return;
+    }
+    auto& module = *getModule();
+    auto table = module.getTable(tableName);
+    if (table->is64()) {
+      assert(ptr->type == Type::i64);
+      ptr->type = Type::i32;
+      Builder builder(module);
+      ptr = builder.makeUnary(UnaryOp::ExtendUInt32, ptr);
+    }
+  }
+
+  void visitTableSize(TableSize* curr) {
+    auto& module = *getModule();
+    auto table = module.getTable(curr->table);
+    if (table->is64()) {
+      auto size = static_cast<Expression*>(curr);
```

Comment:
```suggestion
      auto* size = static_cast<Expression*>(curr);
```

Reply:
Though we more commonly write
```suggestion
      Expression* size = curr;
```
But maybe static_cast is better?